File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-211617-6449/workdir/src/customerimporter/zip_code.py
from abc import abstractmethod

from customerimporter.imported_value import ImportedValue


class ZipCode(ImportedValue):

    # instance creation

    @classmethod
    def for_(cls, a_zip_code):
        return cls._type_that_can_handle(a_zip_code).with_value(a_zip_code)

    # error messages

    @classmethod
    def invalid_type_error_description(cls):
        # the tests expect the same description used for an invalid identification type
        return "Invalid identification type"

    # code

    @abstractmethod
    def code(self):
        pass

    def new_code_if_none(self, a_none_block):
        return a_none_block()

    def old_code_if_none(self, a_none_block):
        return a_none_block()

    # testing

    def is_new(self):
        return False

    def is_old(self):
        return False

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-204311-76559/analysis/own-tests-project/src/main/java/customerimporter/ZipCode.java
package customerimporter;

import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public abstract class ZipCode {

    // Each zip code type knows whether it is the one written in the record, so adding a new type
    // does not change this class more than this list.
    private static final List<ZipCodeCreator> creators = List.of(
        new ZipCodeCreator(OldZipCode::canBeCreatedFrom, OldZipCode::new),
        new ZipCodeCreator(NewZipCode::canBeCreatedFrom, NewZipCode::new));

    // instance creation

    public static ZipCode from(String aZipCode) {
        return creators.stream()
            .filter(aCreator -> aCreator.canCreateFrom(aZipCode))
            .findFirst()
            .orElseThrow(() -> new RuntimeException(invalidZipCodeTypeErrorDescription()))
            .from(aZipCode);
    }

    // error messages

    public static String invalidZipCodeTypeErrorDescription() {
        // Same message as the original code - and as the one the tests expect - even though
        // it talks about an identification type
        return "Invalid identification type";
    }

    // zip code

    public abstract Object value();

    // testing

    public abstract boolean isNew();

    public abstract boolean isOld();

    // values

    public abstract Object newZipCodeIfNone(Supplier<Object> aNoneBlock);

    public abstract Object oldZipCodeIfNone(Supplier<Object> aNoneBlock);

    // instance creation - private

    private record ZipCodeCreator(Predicate<String> condition, Function<String, ZipCode> creation) {

        boolean canCreateFrom(String aZipCode) {
            return condition.test(aZipCode);
        }

        ZipCode from(String aZipCode) {
            return creation.apply(aZipCode);
        }
    }
}
